src/protocol/oprf/mod.rs

```diff
@@ -0,0 +1,330 @@
+
+    // 2. Run computations
+    let y_1 = receive_from_peer(
+        ctx,
+        &OPRFShuffleStep::TransferY1,
+        Direction::Left,
+        batch_size,
+    )
+    .await?;
+
+    let mut y_2: Vec<OPRFShare> = add_single_shares(y_1, z_31).collect();
+    apply_permutation(&pi_31, &mut y_2);
+
+    let mut y_3: Vec<OPRFShare> = add_single_shares(y_2, z_23).collect();
+    apply_permutation(&pi_23, &mut y_3);
+
+    let c_hat_2 = add_single_shares(y_3.iter(), a_hat.iter()).collect::<Vec<_>>();
+    let ((), c_hat_1) = future::try_join(
+        send_to_peer(
+            ctx,
+            &OPRFShuffleStep::TransferCHat,
+            Direction::Left,
+            c_hat_2.clone(),
+        ),
+        receive_from_peer(
+            ctx,
+            &OPRFShuffleStep::TransferCHat,
+            Direction::Left,
+            batch_size,
+        ),
+    )
+    .await?;
+
+    let c_hat = add_single_shares(c_hat_1, c_hat_2);
+    let res = combine_shares(c_hat, a_hat);
+    Ok(res)
+}
+
+// --------------------------------------------------------------------------- //
+
+fn split_shares(
+    input_rows: &[OPRFInputRow],
+    direction: Direction,
+) -> impl Iterator<Item = OPRFShare> + '_ {
+    let f = move |input_row| OPRFShare::from_input_row(input_row, direction);
+    input_rows.iter().map(f)
+}
+
+fn combine_shares<L, R>(l: L, r: R) -> Vec<OPRFInputRow>
+where
+    L: IntoIterator<Item = OPRFShare>,
+    R: IntoIterator<Item = OPRFShare>,
+{
+    l.into_iter()
+        .zip(r)
+        .map(|(l, r)| l.to_input_row(r))
+        .collect::<Vec<_>>()
+}
+
+fn add_single_shares<'i, T, L, R>(l: L, r: R) -> impl Iterator<Item = T::Output> + 'i
+where
+    T: Add,
+    L: IntoIterator<Item = T> + 'i,
+    R: IntoIterator<Item = T> + 'i,
+{
+    l.into_iter().zip(r).map(|(a, b)| a + b)
+}
+
+// --------------------------------------------------------------------------- //
+
+fn generate_random_tables_with_peers<C: Context>(
+    batch_size: u32,
+    narrow_ctx: &C,
+) -> (
+    impl Iterator<Item = OPRFShare> + '_,
+    impl Iterator<Item = OPRFShare> + '_,
+) {
+    let (rng_l, rng_r) = narrow_ctx.prss_rng();
+    let with_left = sample_iter(rng_l).take(batch_size as usize);
+    let with_right = sample_iter(rng_r).take(batch_size as usize);
+    (with_left, with_right)
+}
+
+fn generate_random_table_solo<C>(
+    batch_size: u32,
+    narrow_ctx: &C,
+    peer: Direction,
+) -> impl Iterator<Item = OPRFShare> + '_
+where
+    C: Context,
+{
+    let rngs = narrow_ctx.prss_rng();
+    let rng = match peer {
+        Direction::Left => rngs.0,
+        Direction::Right => rngs.1,
+    };
+
+    sample_iter(rng).take(batch_size as usize)
+}
+
+fn sample_iter<R: Rng>(rng: R) -> impl Iterator<Item = OPRFShare> {
+    rng.sample_iter(Standard)
+}
+
+// ---------------------------- helper communication ------------------------------------ //
+
+async fn send_to_peer<C: Context, I: IntoIterator<Item = OPRFShare>>(
```

Comment:
Do we need separate functions for send and receive here? It seems that ctx.send are simple enough to be used directly. Then you don't need to pass all these inputs. @akoshelev what are the best practices here?

Reply:
we don't have to. If it is common to send and receive all records sequentially, we could add a trait to this module that extends iterators and provides `seq_send` method that takes an iterable of messages and sends them all